### ISSUE TITLE
Node Object Inlining

### ISSUE THREAD

**Issue by fniephaus** [oracle/graal#5044]
### TL;DR

Extend the Truffle DSL with the ability to generate inlined versions of nodes to reduce the memory footprint of GraalVM languages.

### Goals

We plan to introduce a new annotation called `@GenerateInline`, which instructs the Truffle DSL annotation processor to generate an inlinable version of a node. This works analogously to `@GenerateCached` and `@GenerateUncached`, which generate a cached or uncached node version. By default, the DSL does not generate an inlined version of a node. Node inlining provides a simple way to reduce the memory footprint of nodes. In a synthetic example consisting of 17 nodes, node object inlining can reduce the overhead from 236 bytes per instance to 18 bytes. We will provide a detailed tutorial on how to use `@GenerateInline`.

### Non-Goals

- Provide refactorings for language implementers to enable node object inlining automatically
- Disallow customizations, language implementers will also be able to manually implement inlinable nodes
